src/boot/efi/linux.c

```diff
@@ -4,89 +4,63 @@
 #include <efilib.h>
 
 #include "linux.h"
+#include "initrd.h"
 #include "util.h"
 
-#ifdef __i386__
-#define __regparm0__ __attribute__((regparm(0)))
-#else
-#define __regparm0__
-#endif
-
-typedef VOID(*handover_f)(VOID *image, EFI_SYSTEM_TABLE *table, struct boot_params *params) __regparm0__;
-
-static VOID linux_efi_handover(EFI_HANDLE image, struct boot_params *params) {
-        handover_f handover;
-        UINTN start = (UINTN)params->hdr.code32_start;
-
-        assert(params);
-
-#ifdef __x86_64__
-        asm volatile ("cli");
-        start += 512;
-#endif
-        handover = (handover_f)(start + params->hdr.handover_offset);
-        handover(image, ST, params);
-}
-
-EFI_STATUS linux_exec(EFI_HANDLE image,
-                      CHAR8 *cmdline, UINTN cmdline_len,
-                      UINTN linux_addr,
-                      UINTN initrd_addr, UINTN initrd_size) {
-
-        const struct boot_params *image_params;
-        struct boot_params *boot_params;
-        EFI_PHYSICAL_ADDRESS addr;
-        UINT8 setup_sectors;
+EFI_STATUS linux_exec(
+        EFI_HANDLE image,
+        CHAR8* cmdline, UINTN cmdline_len,
+        VOID* linux_buffer, UINTN linux_length,
+        VOID* initrd_buffer, UINTN initrd_size
+) {
         EFI_STATUS err;
+        EFI_HANDLE handle;
+        EFI_LOADED_IMAGE* loaded_image;
 
-        assert(image);
```

Comment:
Please keep asserts and add ones for any new params you add

Reply:
I'll add linux_buffer (initrd might be NULL)
Is it realy impossible for cmdline to be NULL (ie, when no arguments where passed to the executable and no .cmdline section was embedded?), stub.c does not indicate otherwise...

Reply:
Yeah, could be. assert(cmdline || cmdline_len == 0) should do the trick then